accepted/00000-upgrading-bootstrap.md

```diff
@@ -0,0 +1,70 @@
+
+This allows us to cut up the effort required and do at least some of the work gradually. I suspect the last core theme change will still be painful, though. It might be easier to rebuild the existing theme from the base Bootstrap style rather than migrating the existing one, but this is something we can consider after we've simplified the theme.  
+
+## Component and class name changes
+
+This is tied to the previous section, but worth discussing on its own. Numerous component, layout and utility classes have been dropped or changed in both upgrades, e.g. `input-group-addon` was changed in Bootstrap 4 and dropped in 5, so we need to use `input-group-text` and check if it renders correctly etc. The list is pretty long. As discussed in [this ticket](https://github.com/SUSE/spacewalk/issues/18346), it's unlikely that we'd find a tool that can handle all different ways we use Bootstrap (in JSP, JPSF, React, etc), however writing a migration script similar to the Typescript migration might be feasible. For this, we can assemble a list that covers all the breaking changes that affect us by going through the list of changes for 4 and 5, checking whether we use the affected classes, and if so, write down what the upgrade path is.  
+
+## Scripts
+
+There are some scripting changes, e.g. static method names have changed `_foo()` → `foo()`, data properties have changed `data-foo` → `data-bs-foo` etc. This mostly concerns old pages and I would expect most issues here to be straightforward to fix once we get to a state where we can run the test suite on it.
+
+## Summary
+
+In short, the suggested upgrade path can be summarized as follows:  
+
+ - Simplify the existing styles, trying to use standard Bootstrap tooling wherever possible
+ - Gather the list of breaking changes
+ - Write a migration script and/or manually migrate the above list
+ - Migrate existing styles from Less to Sass
+ - Verify the results with Cucumber
+ - Clean up any dangling ends
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Suma is huge and the UI likewise. Last time we touched the theme to match corporate styling for 4.3.0 we broke Many Things and the tail end of layout issues is long. Realistically I would expect this upgrade to require considerable time with QA and testing. In an ideal case, the upgrade would be merged as early in a new version cycle as possible, as to give everyone the maximum amount of time possible to ensure the end result is up to par. 
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+The easiest alternative is to do nothing. We've been shipping Bootstrap 3 along with its known issues for a long time and so far it hasn't blown up in our face. Likewise, our current themes are hard to maintain, but so long as corporate branding doesn't change, there isn't too much to change day-to-day. It is possible to go ahead with the stylesheet simplification etc by itself without upgrading Bootstrap.  
```

Comment:
Would it be possible to patch the XSS vulnerabilities downstream instead? Could it be Red Hat, Debian or some other open source project has done that already?

Reply:
Possible yes, reasonable, I'm inclined to say no. Bootstrap is not a trivial piece of code, as evidenced by these vulnerabilities, and since we don't have anyone who's familiar with their codebase I wouldn't be sure we would be fixing these issues correctly without introducing any new issues. It's possible someone has looked into it, but I doubt they would still be maintaining it at this point — I did a cursory search and could not find anything though.

Reply:
Let's rephrase it this way: reducing and simplifying the stylesheet is a short-mid term goal, upgrading Bootstrap is the mid-long term goal that can come only after the short-mid one. Both are worth to be chased and both would deliver value for developers and customers at the same time. I don't see this as an alternative, but just a split of the Epic task into more specific and defined steps to run over. Estimating the whole RFC to be done would be something for _only the braves_, starting with the simplification described above instead sounds more reasonable :wink: